### PR TITLE
[IR] Add getelementptr nusw and nuw flags

### PR DESCRIPTION
**Pull request by nikic**
This implements the `nusw` and `nuw` flags for `getelementptr` as proposed at https://discourse.llvm.org/t/rfc-add-nusw-and-nuw-flags-for-getelementptr/78672.

The three possible flags are encapsulated in the new `GEPNoWrapFlags` class. Currently this class has a ctor from bool, interpreted as the InBounds flag. This ctor should be removed in the future, as code gets migrated to handle all flags.

There are a few places annotated with `TODO(gep_nowrap)`, where I've had to touch code but opted to not infer or precisely preserve the new flags, so as to keep this as NFC as possible and make sure any changes of that kind get test coverage when they are made.

### CODE REVIEW

**Comment by tschuett** [comment]
Could you please add a TODO here:
https://github.com/llvm/llvm-project/blob/e3f42b02a4129947ca2dd820bfb63ffed83027b7/llvm/lib/CodeGen/MachineInstr.cpp#L565
Thanks.

**Comment by antoniofrighetto** [comment]
Are the TODOs encompassing all the cases? Why we don't want to set the flags in `PHITransAddr` as well?

File: llvm/docs/LangRef.rst
Comment (goldsteinn) [inline llvm/docs/LangRef.rst:11316]: missing '(``add nsw``)' at the end of this?

Comment (goldsteinn) [reply llvm/docs/LangRef.rst:11316]: Also maybe im a bit confused, but should this not be 'pointer index type'?

File: llvm/lib/Bitcode/Reader/BitcodeReader.cpp
Comment (goldsteinn) [reply llvm/lib/Bitcode/Reader/BitcodeReader.cpp:1706]: likewise elsewhere.

File: llvm/test/Assembler/flags.ll
Comment (arsenm) [inline llvm/test/Assembler/flags.ll:397]: Maybe test non-0 AS and vectors?

**Comment by nikic** [comment]
> Are the TODOs encompassing all the cases? Why we don't want to set the flags in `PHITransAddr` as well?

No, the TODOs are only for places where I had to modify code, but the used implementation is obviously non-optimal to minimize initial impact. There are many more places that need to be modified to full propagate the flags everywhere.

File: llvm/docs/LangRef.rst
Comment (nikic) [reply llvm/docs/LangRef.rst]: > missing '(`add nsw`)' at the end of this?

No, because the base is unsigned here. This would be `add nusw` if it existed.

> Also maybe im a bit confused, but should this not be 'pointer index type'?

Yeah, I've updated this and other places to use "pointer index type".

File: llvm/lib/Bitcode/Reader/BitcodeReader.cpp
Comment (nikic) [reply llvm/lib/Bitcode/Reader/BitcodeReader.cpp:1706]: I think using `else if` would be correct here as `setIsInBounds` will also set nusw by itself, but keeping all cases separate seems cleaner?

Otherwise we'll have the following code here...
```
        if (BC->Flags & (1 << bitc::GEP_INBOUNDS))
          cast<GetElementPtrInst>(I)->setIsInBounds();
        else if (BC->Flags & (1 << bitc::GEP_NUSW))
          cast<GetElementPtrInst>(I)->setHasNoUnsignedSignedWrap();
        if (BC->Flags & (1 << bitc::GEP_NUW))
          cast<GetElementPtrInst>(I)->setHasNoUnsignedWrap();
```
...at which point it will probably need a comment to explain why there is only one `else if` there.

**Review by aeubanks — APPROVED** [review]
thanks, I think abstracting out GEPNoWrapFlags is good

File: llvm/include/llvm/IR/GEPNoWrapFlags.h
Comment (goldsteinn) [inline llvm/include/llvm/IR/GEPNoWrapFlags.h:41]: Why do we need this CTOR? Seems to be entirely encapsulated with `inBounds` which is more expressive. Seems a bit non-intuitive to only have 1 bool constructor for a flags class that represents 3...

Comment (nikic) [reply llvm/include/llvm/IR/GEPNoWrapFlags.h:41]: This is a backwards-compatiblity ctor because we're replacing existing `bool IsInBounds` parameters with `GEPNoWrapFlags NW` instead. As the TODO indicates, this ctor will be removed in the future. It avoids having to touch large amounts of code in this PR in sub-optimal ways.

**Comment by goldsteinn** [comment]
LGTM. Wait on some additional approvals to push please.

**Comment by nikic** [comment]
There's already another approval from @aeubanks, so I plan to merge this next Monday if there's no more feedback.

**Pull request merged**
